pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/MaximumPinotConsumptionDelayTracker.java

```diff
@@ -0,0 +1,306 @@
+      if ((newMax == null) || (currentMeasure._delayMilliseconds > newMax._delayMilliseconds)) {
+        newMax = currentMeasure;
+      }
+    }
+    return newMax;
+  }
+
+  /*
+   * This function will read ideal state for the given partition and return true if ideal state reflects the
+   * partition must be hosted by the current server, false otherwise.
+   *
+   * @param partitionId Partition ID that we want to verify is still hosted by this server.
+   */
+  private boolean isPartitionHostedByThisServerPerIdealState(long partitionId) {
+    return _realTimeTableDataManager.isPartitionHostedInThisServer(partitionId);
+  }
+
+  /*
+   * Helper function to be called when we should stop tracking a given partition. Removes the partition from
+   * all our maps, it also updates the maximum if the tracked partition was the previous maximum.
+   *
+   * @param partitionId partition ID which we should stop tracking.
+   */
+  private void removePartitionId(long partitionId) {
+    _partitionToDelaySampleMap.remove(partitionId);
+    // If we are removing a partition we should stop reading its ideal state.
+    _partitionsMarkedForConfirmation.remove(partitionId);
+   }
+
+  // Custom Constructor
+  public MaximumPinotConsumptionDelayTracker(ServerMetrics sm, String tableNameWithType, RealtimeTableDataManager tdm,
+      int maxNumberOfTicksBeforeVerification, int initialTimeThreadDelayMs, int timerThreadTickIntervalMs)
+      throws RuntimeException {
+    _serverMetrics = sm;
+    _tableNameWithType = tableNameWithType;
+    _realTimeTableDataManager = tdm;
+    // TBD handle negative timer values
+    if ((maxNumberOfTicksBeforeVerification < 0) || (initialTimeThreadDelayMs < 0)
+        || (timerThreadTickIntervalMs <= 0)) {
+      throw new RuntimeException("Illegal timer arguments");
+    }
+    _enableAging = true;
+    _maxNumberOfTicksBeforeVerification = maxNumberOfTicksBeforeVerification;
+    _initialTimeThreadDelayMs = initialTimeThreadDelayMs;
+    _timerThreadTickIntervalMs = timerThreadTickIntervalMs;
+    _timerTask = new TrackingTimerTask("TimerTask" + tableNameWithType, this);
+    _timer = new Timer();
+    _timer.schedule(_timerTask, _initialTimeThreadDelayMs, _timerThreadTickIntervalMs);
+    // Install callback metric
+    _serverMetrics.addCallbackTableGaugeIfNeeded(_tableNameWithType, ServerGauge.MAX_PINOT_CONSUMPTION_DELAY_MS,
+        () -> (long) getMaxPinotConsumptionDelay());
+  }
+
+  // Constructor that uses default timeout
+  public MaximumPinotConsumptionDelayTracker(ServerMetrics sm, String tableNameWithType,
+      RealtimeTableDataManager tdm) {
+    this(sm, tableNameWithType, tdm, MAX_NUMBER_OF_TIME_TICKS_BEFORE_VERIFICATION,
+        INITIAL_TIMER_THREAD_DELAY_MS, TIMER_THREAD_TICK_INTERVAL_MS);
+  }
+
+  /**
+   * Use to set or rest the aging of reported values.
+   * @param enableAging true if we want maximum to be aged as per sample time or false if we do not want to age
```

Comment:
when will we NOT enable aging (other than test)?

Reply:
If we decided to drop aging, then I think this might simplify the state you need to keep, as I think you will only need to track the delay at the time the event was measured, and can probably drop the sample time altogether.

Reply:
Marked Visible for testing only